gk/fib.c

```diff
@@ -1254,34 +1273,45 @@ init_grantor_fib_locked(struct ip_prefix *ip_prefix,
 		num_cache_holds++;
 	}
 
-	fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
-	if (fib_id < 0)
-		goto put_ether_cache;
+	if (!prefix_exists) {
+		fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
+		if (fib_id < 0)
+			goto put_ether_cache;
+	}
 
 	if (ip_prefix->addr.proto == RTE_ETHER_TYPE_IPV4)
 		gt_fib = &ltbl->fib_tbl[fib_id];
 	else
 		gt_fib = &ltbl->fib_tbl6[fib_id];
 
-	gt_fib->action = GK_FWD_GRANTOR;
-	gt_fib->u.grantor.set = rte_malloc("gk_fib.grantor.set",
-		sizeof(*(gt_fib->u.grantor.set)) +
-			num_addrs * sizeof(*(gt_fib->u.grantor.set->entries)),
-		0);
-	if (gt_fib->u.grantor.set == NULL) {
+	new_set = rte_malloc("gk_fib.grantor.set",
+		sizeof(*new_set) + num_addrs * sizeof(*(new_set->entries)), 0);
+	if (new_set == NULL) {
 		GK_LOG(ERR, "Could not allocate set of Grantor entries\n");
 		goto init_fib;
```

Comment:
We can ` goto put_ether_cache;` here since we have not written anything to the fib entry yet.

In fact, we should swap the labels `init_fib` and `free_set` in the stack later in the code to reflect the fact that we allocate the memory before doing anything with the fib entry.

Reply:
Use `unlikely()` on the expression `new_set == NULL`.